### PR TITLE
Webatrice: KnownHosts component

### PR DESCRIPTION
**Pull request by seavor**

### CODE REVIEW

File: webclient/src/components/KnownHosts/KnownHosts.tsx [hunk 1]
```diff
@@ -1,85 +1,208 @@
...
+                ) }
+              </MenuItem>
+            ))
+          }
+        </Select>
+      </FormControl>
+
+      <KnownHostDialog
+        isOpen={dialogState.open}
+        host={dialogState.edit}
```
Comment (ZeldaZach) [inline webclient/src/components/KnownHosts/KnownHosts.tsx:200]: might've missed why this is called host = edit

File: webclient/src/containers/Login/Login.tsx [hunk 2]
```diff
@@ -73,13 +75,63 @@ const Login = ({ state, description }: LoginProps) => {
...
+      });
+    }
+  }, ServerTypes.LOGIN_SUCCESSFUL, [hostIdToRemember]);
+
   const showDescription = () => {
     return !isConnected && description?.length;
   };
 
-  const createAccount = () => {
+  function createAccount() {
```
Comment (ZeldaZach) [inline webclient/src/containers/Login/Login.tsx:91]: why the change from functional component?

File: webclient/src/containers/Login/Login.tsx [hunk 3]
```diff
@@ -73,13 +75,63 @@ const Login = ({ state, description }: LoginProps) => {
...
+      ...getHostPort(selectedHost),
+      userName,
+      password
+    };
+
+    if (!password) {
+      options.hashedPassword = hashedPassword;
+    }
+
+    AuthenticationService.login(options as WebSocketConnectOptions);
```
Comment (ZeldaZach) [inline webclient/src/containers/Login/Login.tsx:123]: will this drop the extra field(s) (like hashedPassword) before logging in?

File: webclient/src/containers/Login/Login.tsx [hunk 4]
```diff
@@ -73,13 +75,63 @@ const Login = ({ state, description }: LoginProps) => {
...
+  }, []);
+
+  function updateHost({ selectedHost, userName, hashedPassword, remember }) {
```
Comment (ZeldaZach) [inline webclient/src/containers/Login/Login.tsx:126]: functional vs function? We should be really sticking with one approach

File: webclient/src/dialogs/KnownHostDialog/KnownHostDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+          {handleClose ? (
+            <IconButton onClick={handleClose}>
+              <CloseIcon fontSize='large' />
+            </IconButton>
+          ) : null}
+        </div>
+      </DialogTitle>
+      <DialogContent className='dialog-content'>
+        <Typography className='dialog-content__subtitle' variant='subtitle1'>
+          Adding a new host allows you to connect to different servers. Enter the details below to your host list.
```
Comment (ZeldaZach) [inline webclient/src/dialogs/KnownHostDialog/KnownHostDialog.tsx:47]: "Enter the details below to your host list" is a bit awkward

File: webclient/src/forms/LoginForm/LoginForm.tsx [hunk 6]
```diff
@@ -1,46 +1,137 @@
...
+import { useAutoConnect } from 'hooks';
+import { HostDTO, SettingDTO } from 'services';
+import { FormKey, APP_USER } from 'types';
 
 import './LoginForm.css';
 
-const LoginForm = (props) => {
-  const { dispatch, handleSubmit } = props;
+const PASSWORD_LABEL = 'Password';
+const STORED_PASSWORD_LABEL = '* SAVED *';
```
Comment (ZeldaZach) [inline webclient/src/forms/LoginForm/LoginForm.tsx:17]: How about just "Password Saved"?

File: webclient/src/forms/RegisterForm/RegisterForm.tsx [hunk 7]
```diff
@@ -13,15 +13,16 @@ import './RegisterForm.css';
 const RegisterForm = (props) => {
   const { dispatch, handleSubmit } = props;
 
-  const onHostChange = ({ host, port }) => {
+  const onHostChange: any = ({ host, port }) => {
```
Comment (ZeldaZach) [inline webclient/src/forms/RegisterForm/RegisterForm.tsx:16]: Do we want this type'd to any?

File: webclient/src/hooks/useDebounce.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { useCallback } from 'react';
+import { debounce, DebouncedFunc } from 'lodash';
+
+type UseDebounceType = (...args: any) => any;
+const DEBOUNCE_DELAY = 250;
```
Comment (ZeldaZach) [inline webclient/src/hooks/useDebounce.ts:5]: should have time increment, DEBOUNCE_DELAY_SEC or smth

File: webclient/src/services/DexieService.ts [hunk 9]
```diff
@@ -18,7 +20,11 @@ class DexieService {
   private db: Dexie = new Dexie('Webatrice');
 
   constructor() {
-    this.db.version(1).stores(StoreKeyIndexes);
+    this.db.version(2).stores(StoreKeyIndexes);
```
Comment (ZeldaZach) [inline webclient/src/services/DexieService.ts:23]: Do we include an upgrade path for users who may have an outdated local?

File: webclient/src/store/server/server.interfaces.ts [hunk 10]
```diff
@@ -3,24 +3,26 @@ import { Log, SortBy, User, UserSortField } from 'types';
...
+  userName: string;
+  password: string;
 }
 
 export interface ServerRegisterParams {
   host: string;
   port: string;
-  user: string;
-  pass: string;
-  passAgain: string;
```
Comment (ZeldaZach) [inline webclient/src/store/server/server.interfaces.ts:15]: Are we still checking that passwords match when registering?

File: webclient/src/websocket/commands/SessionCommands.ts [hunk 11]
```diff
@@ -132,35 +136,35 @@ export class SessionCommands {
...
 
     const registerConfig = {
       ...webClient.clientConfig,
-      userName: options.user,
-      password: options.pass,
-      email: options.email,
-      country: options.country,
-      realName: options.realName,
-      clientid: 'webatrice'
+      clientid: guid(),
```
Comment (ZeldaZach) [inline webclient/src/websocket/commands/SessionCommands.ts:162]: Might want to leave some kind of bread crumbs this GUID isn't real, having webatrice in it seemed useful (same for other areas of change)

File: webclient/src/websocket/utils/sanitizeHtml.util.ts [hunk 12]
```diff
@@ -17,13 +17,13 @@ export function sanitizeHtml(msg: string): string {
 }
 
 function enforceTagWhitelist($el: JQuery<HTMLElement>, tags: string): void {
-  $el.find('*').not(tags).each(() => {
+  $el.find('*').not(tags).each(function enforceTag() {
```
Comment (ZeldaZach) [inline webclient/src/websocket/utils/sanitizeHtml.util.ts:20]: What's the change about? Same for the other ones below

File: webclient/tsconfig.json [hunk 13]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "baseUrl": "src",
-    "target": "es5",
+    "target": "es6",
```
Comment (ZeldaZach) [inline webclient/tsconfig.json:4]: 👍🏼

File: webclient/src/components/KnownHosts/KnownHosts.tsx (hunk 1 above)
Comment (seavor) [reply webclient/src/components/KnownHosts/KnownHosts.tsx:200]: `edit` holds the host that's going to be edited (see openEditKnownHostDialog method)

File: webclient/src/containers/Login/Login.tsx (hunk 3 above)
Comment (seavor) [reply webclient/src/containers/Login/Login.tsx:123]: If passwordSalt is supported, the protobuf command will look for a hashedPassword, or hash the password itself if hashedPassword wasnt provided. If its not supposed, it will just use the naked password.

File: webclient/src/containers/Login/Login.tsx (hunk 4 above)
Comment (seavor) [reply webclient/src/containers/Login/Login.tsx:126]: Good catch, i forget why i did this but i dont need to.

File: webclient/src/dialogs/KnownHostDialog/KnownHostDialog.tsx (hunk 5 above)
Comment (seavor) [reply webclient/src/dialogs/KnownHostDialog/KnownHostDialog.tsx:47]: We can work on the language as a follow up.

File: webclient/src/forms/LoginForm/LoginForm.tsx (hunk 6 above)
Comment (seavor) [reply webclient/src/forms/LoginForm/LoginForm.tsx:17]: I was planning to run this by Lee and the group as a follow up.

File: webclient/src/forms/RegisterForm/RegisterForm.tsx (hunk 7 above)
Comment (seavor) [reply webclient/src/forms/RegisterForm/RegisterForm.tsx:16]: yes

File: webclient/src/hooks/useDebounce.ts (hunk 8 above)
Comment (seavor) [reply webclient/src/hooks/useDebounce.ts:5]: Milliseconds is kinda the default standard.

File: webclient/src/services/DexieService.ts (hunk 9 above)
Comment (seavor) [reply webclient/src/services/DexieService.ts:23]: Not yet.

File: webclient/src/containers/Login/Login.tsx (hunk 2 above)
Comment (seavor) [reply webclient/src/containers/Login/Login.tsx:91]: Good catch, forget why i did this, will change back.

File: webclient/src/websocket/utils/sanitizeHtml.util.ts (hunk 12 above)
Comment (seavor) [reply webclient/src/websocket/utils/sanitizeHtml.util.ts:20]: lint rule.

File: webclient/src/store/server/server.interfaces.ts (hunk 10 above)
Comment (seavor) [reply webclient/src/store/server/server.interfaces.ts:15]: Yes, but that will be validated locally in the component, no need to pass the second password up the chain.

File: webclient/src/hooks/useDebounce.ts (hunk 8 above)
Comment (ZeldaZach) [reply webclient/src/hooks/useDebounce.ts:5]: It's dangerous to assume unit defaults

File: webclient/src/services/DexieService.ts (hunk 9 above)
Comment (ZeldaZach) [reply webclient/src/services/DexieService.ts:23]: Should ticket this for later

**Review by ZeldaZach — APPROVED** [review]
Seems good enough for now. A few tweaks to be made in the future. Nice work!

**Review by ZeldaZach — APPROVED** [review]
Tested and now lgtm
